### ISSUE TITLE
[Serve] The API routes for ServeAgent are not available

### ISSUE THREAD

**Issue by VishDev12** [ray-project/ray#34882]
### What happened + What you expected to happen

I'm trying to run: `serve deploy application.yaml`.

First indication of failure was when I tried to `serve submit` a serve configuration file to a Ray cluster. I received:

```
RuntimeError: Version check returned 404. Serve CLI is not supported on the Ray cluster. Please ensure the cluster is running Ray 1.12 or higher.
```

I checked the corresponding location in the code:

https://github.com/ray-project/ray/blob/daa6df020abc992b27f070ba9298584a578d62e5/dashboard/modules/dashboard_sdk.py#L241-L247

And when I printed the `url`, I saw it was `/api/ray/version`. When I curled the dashboard URL with this route, it returned a 404.

I also tested `/api/version` but that worked:

```
{"version": "4", "ray_version": "2.4.0", "ray_commit": "4479f66d4db967d3c9dd0af2572061276ba926ba"}
```

---

Next, I checked  `serve_agent.py` where the failing route was located and realized none of the routes there seemed to work. And I also checked `serve_head.py`, but interestingly, the route there: `/api/serve_head/applications/` worked and I got: `{"deploy_mode": "UNSET", "applications": {}}`.

Now I figured maybe the dashboard process responsible for loading these modules failed, but when I checked the dashboard agent logs, they seemed fine:

```
2023-04-28 13:24:00,118	INFO agent.py:116 -- Parent pid is 1717
2023-04-28 13:24:00,119	INFO agent.py:142 -- Dashboard agent grpc address: 0.0.0.0:57310
2023-04-28 13:24:00,121	INFO utils.py:112 -- Get all modules by type: DashboardAgentModule
2023-04-28 13:24:00,327	INFO utils.py:145 -- Available modules: [<class 'ray.dashboard.modules.event.event_agent.EventAgent'>, <class 'ray.dashboard.modules.healthz.healthz_agent.HealthzAgent'>, <class 'ray.dashboard.modules.job.job_agent.JobAgent'>, <class 'ray.dashboard.modules.log.log_agent.LogAgent'>, <class 'ray.dashboard.modules.log.log_agent.LogAgentV1Grpc'>, <class 'ray.dashboard.modules.reporter.reporter_agent.ReporterAgent'>, <class 'ray.dashboard.modules.runtime_env.runtime_env_agent.RuntimeEnvAgent'>, <class 'ray.dashboard.modules.serve.serve_agent.ServeAgent'>]
2023-04-28 13:24:00,328	INFO agent.py:171 -- Loading DashboardAgentModule: <class 'ray.dashboard.modules.event.event_agent.EventAgent'>
2023-04-28 13:24:00,328	INFO event_agent.py:38 -- Event agent cache buffer size: 10240
2023-04-28 13:24:00,328	INFO agent.py:171 -- Loading DashboardAgentModule: <class 'ray.dashboard.modules.healthz.healthz_agent.HealthzAgent'>
2023-04-28 13:24:00,328	INFO agent.py:171 -- Loading DashboardAgentModule: <class 'ray.dashboard.modules.job.job_agent.JobAgent'>
2023-04-28 13:24:00,328	INFO agent.py:171 -- Loading DashboardAgentModule: <class 'ray.dashboard.modules.log.log_agent.LogAgent'>
2023-04-28 13:24:00,334	INFO agent.py:171 -- Loading DashboardAgentModule: <class 'ray.dashboard.modules.log.log_agent.LogAgentV1Grpc'>
2023-04-28 13:24:00,334	INFO agent.py:171 -- Loading DashboardAgentModule: <class 'ray.dashboard.modules.reporter.reporter_agent.ReporterAgent'>
2023-04-28 13:24:00,337	INFO agent.py:171 -- Loading DashboardAgentModule: <class 'ray.dashboard.modules.runtime_env.runtime_env_agent.RuntimeEnvAgent'>
2023-04-28 13:24:00,338	INFO agent.py:171 -- Loading DashboardAgentModule: <class 'ray.dashboard.modules.serve.serve_agent.ServeAgent'>
2023-04-28 13:24:00,338	INFO agent.py:176 -- Loaded 8 modules.
2023-04-28 13:24:00,342	INFO http_server_agent.py:74 -- Dashboard agent http address: 0.0.0.0:52365
2023-04-28 13:24:00,343	INFO http_server_agent.py:81 -- <ResourceRoute [GET] <PlainResource  /api/local_raylet_healthz> -> <function HealthzAgent.health_check at 0x7f9596379fc0>
2023-04-28 13:24:00,343	INFO http_server_agent.py:81 -- <ResourceRoute [OPTIONS] <PlainResource  /api/local_raylet_healthz> -> <bound method _PreflightHandler._preflight_handler of <aiohttp_cors.cors_config._CorsConfigImpl object at 0x7f95960c1150>>
2023-04-28 13:24:00,343	INFO http_server_agent.py:81 -- <ResourceRoute [POST] <PlainResource  /api/job_agent/jobs/> -> <function JobAgent.submit_job at 0x7f95963d1ab0>
2023-04-28 13:24:00,343	INFO http_server_agent.py:81 -- <ResourceRoute [OPTIONS] <PlainResource  /api/job_agent/jobs/> -> <bound method _PreflightHandler._preflight_handler of <aiohttp_cors.cors_config._CorsConfigImpl object at 0x7f95960c1150>>
2023-04-28 13:24:00,343	INFO http_server_agent.py:81 -- <ResourceRoute [POST] <DynamicResource  /api/job_agent/jobs/{job_or_submission_id}/stop> -> <function JobAgent.stop_job at 0x7f95963d1c60>
2023-04-28 13:24:00,343	INFO http_server_agent.py:81 -- <ResourceRoute [OPTIONS] <DynamicResource  /api/job_agent/jobs/{job_or_submission_id}/stop> -> <bound method _PreflightHandler._preflight_handler of <aiohttp_cors.cors_config._CorsConfigImpl object at 0x7f95960c1150>>
2023-04-28 13:24:00,343	INFO http_server_agent.py:81 -- <ResourceRoute [DELETE] <DynamicResource  /api/job_agent/jobs/{job_or_submission_id}> -> <function JobAgent.delete_job at 0x7f95963d1e10>
2023-04-28 13:24:00,343	INFO http_server_agent.py:81 -- <ResourceRoute [OPTIONS] <DynamicResource  /api/job_agent/jobs/{job_or_submission_id}> -> <bound method _PreflightHandler._preflight_handler of <aiohttp_cors.cors_config._CorsConfigImpl object at 0x7f95960c1150>>
2023-04-28 13:24:00,343	INFO http_server_agent.py:81 -- <ResourceRoute [GET] <DynamicResource  /api/job_agent/jobs/{job_or_submission_id}/logs> -> <function JobAgent.get_job_logs at 0x7f95963d1fc0>
2023-04-28 13:24:00,343	INFO http_server_agent.py:81 -- <ResourceRoute [OPTIONS] <DynamicResource  /api/job_agent/jobs/{job_or_submission_id}/logs> -> <bound method _PreflightHandler._preflight_handler of <aiohttp_cors.cors_config._CorsConfigImpl object at 0x7f95960c1150>>
2023-04-28 13:24:00,343	INFO http_server_agent.py:81 -- <ResourceRoute [GET] <DynamicResource  /api/job_agent/jobs/{job_or_submission_id}/logs/tail> -> <function JobAgent.tail_job_logs at 0x7f95963d2170>
2023-04-28 13:24:00,343	INFO http_server_agent.py:81 -- <ResourceRoute [OPTIONS] <DynamicResource  /api/job_agent/jobs/{job_or_submission_id}/logs/tail> -> <bound method _PreflightHandler._preflight_handler of <aiohttp_cors.cors_config._CorsConfigImpl object at 0x7f95960c1150>>
2023-04-28 13:24:00,343	INFO http_server_agent.py:81 -- <ResourceRoute [GET] <PlainResource  /api/ray/version> -> <function ServeAgent.get_version at 0x7f95960155a0>
2023-04-28 13:24:00,343	INFO http_server_agent.py:81 -- <ResourceRoute [OPTIONS] <PlainResource  /api/ray/version> -> <bound method _PreflightHandler._preflight_handler of <aiohttp_cors.cors_config._CorsConfigImpl object at 0x7f95960c1150>>
2023-04-28 13:24:00,344	INFO http_server_agent.py:81 -- <ResourceRoute [GET] <PlainResource  /api/serve/deployments/> -> <function ServeAgent.get_all_deployments at 0x7f9596015630>
2023-04-28 13:24:00,344	INFO http_server_agent.py:81 -- <ResourceRoute [OPTIONS] <PlainResource  /api/serve/deployments/> -> <bound method _PreflightHandler._preflight_handler of <aiohttp_cors.cors_config._CorsConfigImpl object at 0x7f95960c1150>>
2023-04-28 13:24:00,344	INFO http_server_agent.py:81 -- <ResourceRoute [GET] <PlainResource  /api/serve/applications/> -> <function ServeAgent.get_serve_instance_details at 0x7f95960157e0>
2023-04-28 13:24:00,344	INFO http_server_agent.py:81 -- <ResourceRoute [OPTIONS] <PlainResource  /api/serve/applications/> -> <bound method _PreflightHandler._preflight_handler of <aiohttp_cors.cors_config._CorsConfigImpl object at 0x7f95960c1150>>
2023-04-28 13:24:00,344	INFO http_server_agent.py:81 -- <ResourceRoute [GET] <PlainResource  /api/serve/deployments/status> -> <function ServeAgent.get_all_deployment_statuses at 0x7f9596015990>
2023-04-28 13:24:00,344	INFO http_server_agent.py:81 -- <ResourceRoute [OPTIONS] <PlainResource  /api/serve/deployments/status> -> <bound method _PreflightHandler._preflight_handler of <aiohttp_cors.cors_config._CorsConfigImpl object at 0x7f95960c1150>>
2023-04-28 13:24:00,344	INFO http_server_agent.py:81 -- <ResourceRoute [DELETE] <PlainResource  /api/serve/deployments/> -> <function ServeAgent.delete_serve_application at 0x7f9596015b40>
2023-04-28 13:24:00,344	INFO http_server_agent.py:81 -- <ResourceRoute [OPTIONS] <PlainResource  /api/serve/deployments/> -> <bound method _PreflightHandler._preflight_handler of <aiohttp_cors.cors_config._CorsConfigImpl object at 0x7f95960c1150>>
2023-04-28 13:24:00,344	INFO http_server_agent.py:81 -- <ResourceRoute [DELETE] <PlainResource  /api/serve/applications/> -> <function ServeAgent.delete_serve_applications at 0x7f9596015cf0>
2023-04-28 13:24:00,344	INFO http_server_agent.py:81 -- <ResourceRoute [OPTIONS] <PlainResource  /api/serve/applications/> -> <bound method _PreflightHandler._preflight_handler of <aiohttp_cors.cors_config._CorsConfigImpl object at 0x7f95960c1150>>
2023-04-28 13:24:00,344	INFO http_server_agent.py:81 -- <ResourceRoute [PUT] <PlainResource  /api/serve/deployments/> -> <function ServeAgent.put_all_deployments at 0x7f9596015ea0>
2023-04-28 13:24:00,344	INFO http_server_agent.py:81 -- <ResourceRoute [OPTIONS] <PlainResource  /api/serve/deployments/> -> <bound method _PreflightHandler._preflight_handler of <aiohttp_cors.cors_config._CorsConfigImpl object at 0x7f95960c1150>>
2023-04-28 13:24:00,344	INFO http_server_agent.py:81 -- <ResourceRoute [PUT] <PlainResource  /api/serve/applications/> -> <function ServeAgent.put_all_applications at 0x7f9596016050>
2023-04-28 13:24:00,344	INFO http_server_agent.py:81 -- <ResourceRoute [OPTIONS] <PlainResource  /api/serve/applications/> -> <bound method _PreflightHandler._preflight_handler of <aiohttp_cors.cors_config._CorsConfigImpl object at 0x7f95960c1150>>
2023-04-28 13:24:00,344	INFO http_server_agent.py:81 -- <ResourceRoute [GET] <StaticResource  /logs -> PosixPath('/tmp/ray/session_2023-04-28_13-23-53_318399_1616/logs')> -> <bound method StaticResource._handle of <StaticResource  /logs -> PosixPath('/tmp/ray/session_2023-04-28_13-23-53_318399_1616/logs')>>
2023-04-28 13:24:00,345	INFO http_server_agent.py:81 -- <ResourceRoute [OPTIONS] <StaticResource  /logs -> PosixPath('/tmp/ray/session_2023-04-28_13-23-53_318399_1616/logs')> -> <bound method _PreflightHandler._preflight_handler of <aiohttp_cors.cors_config._CorsConfigImpl object at 0x7f95960c1150>>
2023-04-28 13:24:00,345	INFO http_server_agent.py:82 -- Registered 30 routes.
2023-04-28 13:24:00,350	INFO event_agent.py:56 -- Report events to 172.31.29.146:33147
2023-04-28 13:24:00,350	INFO event_utils.py:132 -- Monitor events logs modified after 1682686440.2548954 on /tmp/ray/session_2023-04-28_13-23-53_318399_1616/logs/events, the source types are all.
2023-04-28 13:24:00,576	INFO event_utils.py:170 -- Found new event log file: /tmp/ray/session_2023-04-28_13-23-53_318399_1616/logs/events/event_AUTOSCALER.log
2023-04-29 00:46:59,415	INFO optional_utils.py:264 -- Connecting to ray with address=172.31.29.146:6379
2023-04-29 00:46:59,415	INFO worker.py:1432 -- Connecting to existing Ray cluster at address: 172.31.29.146:6379...
2023-04-29 00:46:59,421	INFO worker.py:1616 -- Connected to Ray cluster. View the dashboard at [1m[32mhttp://172.31.29.146:8265 [39m[22m
```

So long story short, I have no idea what's happening. I also faced the same error in 2.3.1.

Would really appreciate some help!

### Versions / Dependencies

Ray: 2.3.1 and 2.4.0
Python: 3.10.6

### Reproduction script

NA

### Issue Severity

High: It blocks me from completing my task.

**Comment by catqaq** [comment]
same error:
python3.10/site-packages/ray/dashboard/modules/dashboard_sdk.py", line 264, in _check_connection_and_version_with_url
    raise RuntimeError(
RuntimeError: Version check returned 404. Jobs API is not supported on the Ray cluster. Please ensure the cluster is running Ray 1.9 or higher.

**Comment by chris-aeviator** [comment]
facing this in P3.8, ray 2.7.1.

**Comment by johntaylor-cell** [comment]
Closing as obsolete. The dashboard Serve agent this reports against no longer exists: `python/ray/dashboard/modules/serve/serve_agent.py` has been removed, and the Serve REST API is served from the dashboard head on port 8265 (see #40048).

If you are still seeing a 404 on a Serve REST route, please open a fresh issue against the dashboard-head endpoint with the exact path.
